app/src/main/java/com/google/android/samples/socialite/ui/camera/CameraViewModel.kt

```diff
@@ -343,14 +491,73 @@ class CameraViewModel @Inject constructor(
+        camera.cameraControl.setZoomRatio(finalScale)
+    }
+
+    inner class SelfieSegmentationAnalyzer : ImageAnalysis.Analyzer {
+
+        val options = SelfieSegmenterOptions.Builder()
+            .setDetectorMode(SelfieSegmenterOptions.STREAM_MODE)
+            .build()
+        val selfieSegmenter = Segmentation.getClient(options)
+        var isAnalyzing = false
+
+        @RequiresApi(Build.VERSION_CODES.Q)
+        @ExperimentalGetImage
+        override fun analyze(imageProxy: ImageProxy) {
+            val mediaImage = imageProxy.image
+            if (mediaImage != null && !isAnalyzing) {
+                isAnalyzing = true
+                val image = InputImage.fromMediaImage(mediaImage, imageProxy.imageInfo.rotationDegrees)
+                selfieSegmenter.process(image)
+                    .addOnSuccessListener { results ->
+                        // Get foreground probabilities for each pixel. Since ML Kit returns this
+                        // in a byte buffer with each 4 bytes representing a float, convert it to
+                        // a FloatArray for easier use.
+                        val maskProbabilities = FloatArray(results.buffer.capacity() / 4)
+                        results.buffer.asFloatBuffer().get(maskProbabilities)
+
+                        // If the pixel foreground likelihood is above the threshold, assign it
+                        // with a color, otherwise assign it transparent so it will be masked out.
+                        val maskColors = maskProbabilities.map { point ->
+                            if (point > backgroundRemovalThreshold) {
+                                Color.WHITE
+                            } else {
+                                Color.TRANSPARENT
+                            }
+                        }.toIntArray()
+
+                        bitmap = imageProxy.toBitmap()
+                        mask = Bitmap.createBitmap(maskColors, mediaImage.height, mediaImage.width, Bitmap.Config.ARGB_8888)
+
+                        // Create final bitmap and mask, performing transformations so that they match.
+                        val matrix = Matrix()
+                        matrix.preScale(-1f, 1f)
+                        bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, false)
+
+                        matrix.preRotate(90F)
+                        mask = Bitmap.createBitmap(mask, 0, 0, mask.width, mask.height, matrix, false)
+
+                        // Final cleanup. Close imageProxy for next analysis frame.
+                        imageProxy.close()
+                        isAnalyzing = false
+                    }
+                    .addOnFailureListener { e ->
+                        // Task failed with an exception. Close imageProxy for next analysis frame.
+                        imageProxy.close()
+                        isAnalyzing = false
+                    }
```

Comment:
These can be combined into:
```
.addOnCompleteListener {
    imageProxy.close()
    isAnalyzing = false
}
```

The GMS Task API uses a queue for added listeners. So this will be called at the end if you add it after your OnSuccessListener.

Reply:
Done